pilot/pkg/proxy/envoy/v2/mesh_discovery.go

```diff
@@ -0,0 +1,219 @@
+	var grpcOptions []grpc.ServerOption
+
+	var interceptors []grpc.UnaryServerInterceptor
+
+	// TODO: log request interceptor if debug enabled.
+
+	// setup server prometheus monitoring (as final interceptor in chain)
+	interceptors = append(interceptors, grpc_prometheus.UnaryServerInterceptor)
+	grpc_prometheus.EnableHandlingTimeHistogram()
+
+	grpcOptions = append(grpcOptions, grpc.UnaryInterceptor(grpc_middleware.ChainUnaryServer(interceptors...)))
+
+	// get the grpc server wired up
+	grpc.EnableTracing = true
+
+	grpcServer := grpc.NewServer(grpcOptions...)
+	out := &DiscoveryServer{mesh: mesh, grpcServer: grpcServer}
+	xdsapi.RegisterEndpointDiscoveryServiceServer(out.grpcServer, out)
+	return out
+}
+
+// Start opens up a listening port for serving gRPC requests and starts up the underlying gRPC server.
+// This method is non-blocking.
+func (s *DiscoveryServer) Start() {
+	lis, err := net.Listen("tcp", ":"+gRPCPort)
+	if err != nil {
+		log.Errorf("mesh discovery server failed to listen on port %q: %v", gRPCPort, err)
+		return
+	}
+	// Serve is a blocking call.
+	go func(gRPCServer *grpc.Server, lis net.Listener) {
+		log.Infof("attempting to start mesh discovery server on port %q", gRPCPort)
+		if err := gRPCServer.Serve(lis); err != nil {
+			log.Errorf("mesh discovery server failed to serve on port %q: %v", gRPCPort, err)
+		}
+	}(s.grpcServer, lis)
+}
+
+/***************************  Mesh EDS Implementation **********************************/
+
+// StreamEndpoints implements xdsapi.EndpointDiscoveryServiceServer.StreamEndpoints().
+func (s *DiscoveryServer) StreamEndpoints(stream xdsapi.EndpointDiscoveryService_StreamEndpointsServer) error {
+	ticker := time.NewTicker(responseTickDuration)
+	peerInfo, ok := peer.FromContext(stream.Context())
+	peerAddr := "Unknown peer address"
+	if ok {
+		peerAddr = peerInfo.Addr.String()
+	}
+	defer ticker.Stop()
+	var discReq xdsapi.DiscoveryRequest
+	var receiveError error
+	var clusters []string
+	initialRequest := true
+	reqChannel := make(chan xdsapi.DiscoveryRequest, 1)
+	go func() {
+		defer close(reqChannel)
+		for {
+			req, err := stream.Recv()
+			if err != nil {
+				if status.Code(err) == codes.Canceled || err == io.EOF {
+					return
+				}
+				receiveError = err
+				log.Errorf("request loop for EDS for client %q terminated with errors %v", peerAddr, err)
+				return
+			}
+			reqChannel <- *req
+		}
+	}()
+	for {
+		// Block until either a request is received or the ticker ticks
+		select {
+		case discReq, ok = <-reqChannel:
+			if !ok {
+				return receiveError
+			}
+			if !initialRequest {
+				// Given that Pilot holds an eventually consistent data model, Pilot ignores any acknowledgements
+				// from Envoy, whether they indicate ack success or ack failure of Pilot's previous responses.
+				// Only the first request is actually horored and processed. Pilot drains all other requests from this stream.
+				continue
```

Comment:
Can we add a debug statement dumping the version info and nonce from the Envoy response? I suspect we'll want that for debugging down the line.